### PR TITLE
BUGFIX: Its possible to send a supported group that is not supported.

### PR DESCRIPTION
**Pull request by anhu**
This change fixes that.
This also fixes https://github.com/wolfSSL/wolfssl/issues/4343

### CODE REVIEW

File: src/tls.c [hunk 1]
```diff
@@ -4717,8 +4717,13 @@ int TLSX_UseSupportedCurve(TLSX** extensions, word16 name, void* heap)
     SupportedCurve* curve = NULL;
     int ret;
 
-    if (extensions == NULL)
+    if (extensions == NULL) {
         return BAD_FUNC_ARG;
+    }
+
+    if (! TLSX_KeyShare_IsSupported(name)) {
```
Comment (dgarske) [inline src/tls.c:4724]: Check the pre-def static at tls.c66. Currently only for TLS v1.3 and HAVE_SUPPORTED_CURVES. This should be expanded. You can reproduce with `./configure --disable-tls13`.

```
src/tls.c:4724:11: error: implicit declaration of function 'TLSX_KeyShare_IsSupported' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    if (! TLSX_KeyShare_IsSupported(name)) {
          ^
```

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 1 above)
Comment (anhu) [reply src/tls.c:4724]: ah...Thats why jenkins is being so noisy! Thanks.

**Comment by anhu** [comment]
`Enable Options part C ` failure reproduced . 

Looking into it.

File: src/tls.c [hunk 2]
```diff
@@ -8621,7 +8701,7 @@ int TLSX_KeyShare_Establish(WOLFSSL *ssl, int* doHelloRetry)
 
         /* Check consistency now - extensions in any order. */
         if (!TLSX_SupportedGroups_Find(ssl, clientKSE->group))
-            return BAD_KEY_SHARE_DATA;
+            continue;
```
Comment (dgarske) [inline src/tls.c:8704]: Can you explain or add comment for changing `BAD_KEY_SHARE_DATA` -> `continue`?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 2 above)
Comment (anhu) [reply src/tls.c:8704]: hmmm....well, I noticed a test against openssl that was not passing. OpenSSL was sending supported groups with the first one being X25519 and that was causing this to return `BAD_KEY_SHARE_DATA`; it wasn't looking to the next algorithm in the list.  This allows it to look to the next algo in the list. 

So, the answer above probably isn't what you're looking for.  You probably want to understand what it is about my other changes that triggered the `BAD_KEY_SHARE_DATA`.  I will do a bit more digging.

Comment (anhu) [reply src/tls.c:8704]: Now that begs, the question, why was `X25519` in that list. I'll go dig some more.

Comment (anhu) [reply src/tls.c:8704]: Cha-Ching! In `DoTls13ClientHello()` (where the server handles client hello message) we eventually call `TLSX_SupportedCurve_Parse()` which then calls `TLSX_UseSupportedCurve()`.  Before my change, it never actually checked for whether we supported the client's curve.  Now that we check, X25519 doesn't get put in the list. 

Wow....that was pretty convoluted.

**Review by SparkiDev — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Comment by anhu** [comment]
@dgarske , Thanks!!  I'm wondering if there are more breakages that were unforeseen.

Can you think of anything?

**Comment by dgarske** [comment]
> @dgarske , Thanks!! I'm wondering if there are more breakages that were unforeseen.
> 
> Can you think of anything?

As discussed here are some items to consider (documenting our discussion):

1) Document any changes in API behavior in `doc/dox_comments/header_files`.

You stated "Before that change, you could set supported groups to any group you wanted (note the changes I had to do for unit tests).  So if someone was doing something wrong like setting curve25519 when it wasn't supported then the connection still wouldn't have worked.  But I think the error behaviour now will look different."

2) If any TLS protocol behaviors changed we need to test and possibly document those changes.

You stated: "I went through a long explanation in the PR about changing an error return to `continue;`. That could potentially change some behaviours during a connection. We're actually checking whether we support it, but before we weren't. If we didn't support, then it would error out somewhere else. So it's not like the connection would have succeeded."

**Pull request merged**